### DISCUSSION TITLE
Offline Install

### DISCUSSION THREAD

**Discussion by wpalaceJHA** [dotnet/interactive#3102] [Q&A] [answered]
Can anyone guide me on why I cannot seem to get dotnet interactive installed by staging the nuget file locally? The file definitely exists in the location I specified, and when I crack it open and inspecting the .nuspec file, the metadata looks to match. What am I missing?

![image](https://github.com/dotnet/interactive/assets/124414767/6365b724-40a3-48c4-b365-0f9bce289077)

**Comment by jonsequitur** [comment, marked as answer]
The `--add-source` option should specify a path to a folder rather than the package.

**Reply by wpalaceJHA** [reply]
Thank you! This worked and I've successfully installed dotnet-interactive; however, VSCode's polyglot notebook extension doesn't seem to recognize it and still tries to reach out over the internet to install it. Do you have any advice on how to get past this?

**Reply by jonsequitur** [reply]
Polyglot Notebooks installs its own copy of a specific version of `dotnet-interactive` but it's not opinionated about which feed to get it from. Assuming you have the correct version's package, you could enable offline installation of that package by modifying your user-level nuget.config to remove remote sources (for example nuget.org), and add a local folder path containing the needed packages.

**Reply by wpalaceJHA** [reply]
Excellent! Thanks again.

**Reply by wpalaceJHA** [reply]
I have one more related question. The above comments were very helpful in getting polyglot notebooks working offline, but now that I have it working, I need to add a nuget from within a notebook, and despite having tried changing both my local nuget.config and attempting to use the #i magic command, it is still trying to traverse to nuget.org and failing after many minutes. Do I have something incorrect in the #i or #r magic commands, or could this be a bug?
![image](https://github.com/dotnet/interactive/assets/124414767/e40e026e-7bd0-4a9a-bd4e-dda1ff2a3356)

**Reply by jonsequitur** [reply]
This isn't something that the notebook or .NET Interactive control. 

Do you have an entry for nuget.org in your user-level nuget.config? You would need to remove that by deleting it or adding a `<clear />` element.

**Reply by wpalaceJHA** [reply]
I think I'm giving up on trying to install the SQL kernel on an offline machine, unless someone has an easy button. I'd like to suggest there be a version of Microsoft.DotNet.Interactive.SqlServer released with all its dependencies included. Here is what I attempted: 

On local computer, download the nuget CLI, the SQLkernel nuget, along with all of its dependencies
	1. Download the latest nuget.exe from nuget.org. 
	There is no need to install anything, just place the executable in a folder, such as C:\Tools
	2. cd to the folder and run the following command
	3. ./nuget install Microsoft.DotNet.Interactive.SqlServer -prerelease -recurse -OutputDirectory C:\Temp\NuGet
	4. Compress-Archive C:\Temp\NuGet -Destination sqlinteractive.zip

On target machine
	1. copy nuget.exe
	2. prevent polyglot notebooks from going over the internet to get the nuget with the following commands
	./nuget sources remove -name nuget.org
	./nuget add source C:\Temp\NuGet
	Manually clear the cache by navigating to ~/.packagemanagement/nuget/cache and removing all the files
	3. Extract compressed files to C:\Temp\Nuget
	4. Move all nuget files into the base repo folder:
Get-ChildItem C:\Temp\Nuget -Directory | % {
    Get-ChildItem $_.FullName | ? {$_.Name -match ".nupkg$"} | % {
        Copy-Item $_.FullName -Destination "C:\Temp\NuGet\${$_.Name}"
    }
}

	5. In a notebook, run the magic command 
	#r "nuget: Microsoft.DotNet.Interactive.SqlServer, 1.0.0-beta.23461.4"
	Also tried 
	#r "nuget: Microsoft.DotNet.Interactive.SqlServer, *-*"

There were still dependency issues after doing all this, so I attempted to download the missing ones manually
./nuget install Microsoft.Data.SqlClient.SNI.runtime -version 7.0.304 -prerelease -OutputDirectory C:\Temp\NuGet2
./nuget install Microsoft.SqlServer.Server -prerelease -OutputDirectory C:\Temp\NuGet2
./nuget install System.Runtime.Caching -prerelease -OutputDirectory C:\Temp\NuGet2
./nuget install System.Resources.ResourceManager -prerelease -OutputDirectory C:\Temp\NuGet2
./nuget install System.Security.Cryptography.Cng -prerelease -OutputDirectory C:\Temp\NuGet2
./nuget install System.Security.Cryptography.ProtectedData -prerelease -OutputDirectory C:\Temp\NuGet2
./nuget install Microsoft.NETCore.Targets -prerelease -OutputDirectory C:\Temp\NuGet2
./nuget install System.Text.Encoding -prerelease -OutputDirectory C:\Temp\NuGet2
./nuget install System.Threading.Tasks -prerelease -OutputDirectory C:\Temp\NuGet2
./nuget install System.Security.Cryptography.Pkcs -prerelease -OutputDirectory C:\Temp\NuGet2
./nuget install System.Reflection -prerelease -OutputDirectory C:\Temp\NuGet2
./nuget install NETStandard.Library -prerelease -OutputDirectory C:\Temp\NuGet2
./nuget install Microsoft.NETCore.Platforms -prerelease -OutputDirectory C:\Temp\NuGet2
./nuget install System.Windows.Extensions -prerelease -OutputDirectory C:\Temp\NuGet2

After doing this, there were many more dependency issues popping up, so I gave up here. This is not a showstopper for anything I am trying to do because I can always wrap SQL in PowerShell, but it is not as tidy and I will miss the intellisense when writing SQL. 

![image](https://github.com/dotnet/interactive/assets/124414767/5e2f20d4-1c36-409e-a241-e88372e41d82)
